### PR TITLE
Ep/ Rename Spec to AffineScheme #3345

### LINKED ISSUE #3345: Rename Spec to AffineScheme
We have the types `Spec` and `ProjectiveScheme`. To make the names consistent, we will change `Spec` into `AffineScheme`. The functions to construct affine and projective schemes will still be called `spec` and `proj`. This completes one bullet point in [Schemes Meta Issue #3202](https://github.com/oscar-system/Oscar.jl/issues/3202).

For consistency, we need to also change many related types and functions like `AbsSpec` and `SpecMor`. See https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1961313435 for the list of planned changes.

### PR DESCRIPTION
**Pull request by paemurru**
Resolves #3345

### CODE REVIEW

File: experimental/QuadFormAndIsom/test/setup_tests.jl [hunk 1]
```diff
@@ -1 +0,0 @@
-../../../test/Serialization/setup_tests.jl
```
Comment (joschmitt) [inline experimental/QuadFormAndIsom/test/setup_tests.jl]: This file is actually a symbolic to `../../../test/Serialization/setup_tests.jl`. I assume it is a copy now because you used `sed`. (At least that's what happened to me in the past.)

Comment (paemurru) [reply experimental/QuadFormAndIsom/test/setup_tests.jl]: Fixed. I don't know if there are other symbolic links ruined

File: experimental/LieAlgebras/src/LieAlgebraModuleHom.jl [hunk 2]
```diff
@@ -403,7 +403,7 @@ function canonical_projection(V::LieAlgebraModule, i::Int)
   @req fl "Module must be a direct sum"
   @req 1 <= i <= length(Vs) "Index out of bound"
   j = sum(dim(Vs[l]) for l in 1:(i - 1); init=0)
-  proj = hom(
+  projection = hom(
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/LieAlgebraModuleHom.jl]: What is the purpose of this change and the similar ones in parts of the code that have nothing to do with schemes? I don't really see why they are necessary.

Comment (paemurru) [reply experimental/LieAlgebras/src/LieAlgebraModuleHom.jl]: https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788

Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraModuleHom.jl]: I am with @thofma here: https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950237240

**Review by thofma — DISMISSED** [review]
Variables can be called whatever the original author wants (unless of course it results now in a conflict). This should not be changed here.

**Comment by lgoettgens** [comment]
I think this can be achieved by partially reverting https://github.com/oscar-system/Oscar.jl/pull/3425/commits/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc

**Comment by paemurru** [comment]
@thofma @lgoettgens. The changes in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc are mostly about how to name the maps returned by direct sums and products. In algebraic geometry, naming the returned projection map `proj` conflicts with the homogeneous spectrum. In areas that have connections to algebraic geometry, it is more subjective where to draw the line, where to keep `proj` for internal variable names and where not.

In general, I believe there is value in having consistent naming for variables returned in direct sums and products, so I mostly went with `inj` and `pr`, as that was already common.

OSCAR project leader Michael Joswig encouraged not using `proj` outside of the algebraic geometry meaning, see https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788. Unless you can quickly convince one of the OSCAR project leaders why your point of view is better, I don't think this justifies blocking the merging, especially given our time constraints. If you wish to later change some of the internal variable names back to `proj` in a different pull request, you are welcome to, it is a trivial matter.

**Comment by fieker** [comment]
I as an author name variables in my code. No one else. I as well as several others would be extremely pissed if anyone, for any reason but syntax errors, changes purely internal names. Sorry

**Comment by thofma** [comment]
I think there was a misunderstanding in https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788, which was about functions called `proj` in the global Oscar namespace (which is why there is a reference to "be[ing] globally consistent") and not local variable names. But there does not exist such a function, so there is nothing to adjust.

**Comment by paemurru** [comment]
I think that by "globally consistent" they meant "everywhere consistent" in https://github.com/oscar-system/Oscar.jl/issues/3345#issuecomment-1950227788

**Comment by paemurru** [comment]
If there is a particular change you wish to make regarding the renaming, please make it explicit and I am happy to make the change

**Comment by lgoettgens** [comment]
> If there is a particular change you wish to make regarding the renaming, please make it explicit and I am happy to make the change

As I already wrote earlier, I think this is about everything in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc (but the very few changes regarding to `projectivization` in it). Please correct me if I am wrong @fieker @thofma.

**Comment by paemurru** [comment]
Do I understand correctly that you wish to revert the changes to variables named `proj` in the sense of a projection in local scope in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc, regardless of whether they appear in group theory, algebra or algebraic geometry, regardless of whether they could be confused with the homogeneous spectrum?

**Comment by fieker** [comment]
There is no reason that I can see why an internal variable change is warranted.
If there is a convincing reason, I'd like learn it

**Comment by paemurru** [comment]
@fieker Easier to maintain code, find errors in code, collaborate with others. The same reason why style guides exist in most programming languages, why linters and automatic code formatters exist. The same reason why all code, including internal functions and variables in local scope, are usually formatted according to the style guide.

In this particular case, it is a common best practice not name a variable after functions or variables in global scope (see [StackOverflow](https://stackoverflow.com/a/25000042)). It is something that linters warn against (I haven't found any currently maintained Julia linter, but here is an example from [Pylint](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/redefined-outer-name.html)).

**Comment by paemurru** [comment]
Just to reiterate, I am happy to change the code as you suggest. I just want to make sure I understand what it is you wish to change, as it seems strange to me to use ´proj´ in algebraic geometry for projection morphisms where it can easily be confused with the homogeneous spectrum. It seems strange to me to go against standard best practices and to go against an OSCAR project leader's recommendation.

This renaming issue has been going on for weeks, with different people having different points of view, going back and forth on what or how to rename and with me having to accommodate to these new decisions. I am happy to make the changes, I just hope that I don't need to undo these changes again.

**Comment by thofma** [comment]
> As I already wrote earlier, I think this is about everything in [cb13aad](https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc) (but the very few changes regarding to `projectivization` in it).

**Comment by paemurru** [comment]
Undid all the changes in https://github.com/oscar-system/Oscar.jl/commit/cb13aad4d1fd142c6da5d4274d4f27746e1adcdc except for `proj` -> `projectivization`, as requested by @thofma. This includes changes in files not relating to schemes as well as changes in `Schemes`, changes in documentation and changes that had been approved by the original author in private communication.

**Comment by thofma** [comment]
Thanks! Maybe @simonbrandhorst could give a quick glance and approve the changes here?

**Comment by fingolfin** [comment]
@paemurru since you keep referring to "an OSCAR project leader's recommendation", let me point out that @fieker and @thofma also are OSCAR project leaders and as such I find it odd to draw that card against them... You can of course request that "the OSCAR leaders" sit together and give a final verdict before you change anything, so that you don't end up changing things back and forth (a quite reasonable request).

File: src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Attributes.jl [hunk 3]
```diff
@@ -485,7 +485,7 @@ homogeneous_coordinate_ring(P::ProjectiveScheme) = P.S
 ### type getters
 projective_scheme_type(A::T) where {T<:AbstractAlgebra.Ring} = projective_scheme_type(typeof(A))
 projective_scheme_type(::Type{T}) where {T<:AbstractAlgebra.Ring} =
-ProjectiveScheme{T, mpoly_dec_ring_type(mpoly_ring_type(T))}
+proj{T, mpoly_dec_ring_type(mpoly_ring_type(T))}
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Attributes.jl]: This is a type getter and should therefore return the type.
```suggestion
ProjectiveScheme{T, mpoly_dec_ring_type(mpoly_ring_type(T))}
```

File: src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Constructors.jl [hunk 4]
```diff
@@ -2,13 +2,26 @@
 # Lower case constructors
 ################################################################################
 
-projective_scheme(S::MPolyDecRing) = ProjectiveScheme(S)
+function proj(S::MPolyDecRing)
+  is_standard_graded(S) || error("ring must be standard graded")
+  return ProjectiveScheme(S)
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Constructors.jl]: It seems to me that the better place for this input check would be the actual constructor
`ProjectiveScheme`? (however this need not delay merging this pull request.)

File: src/Groups/directproducts.jl [hunk 5]
```diff
@@ -136,7 +136,7 @@ end
 Return the direct product of `n` copies of `G` as group of type `T`.
 
 The keyword argument `morphisms` is `false` by default. If it is set `true`, then
-the output is a triple (`G`, `emb`, `proj`), where `emb` and `proj` are the
+the output is a triple (`G`, `emb`, `proj`), where `emb` and `pr` are the
```
Comment (simonbrandhorst) [inline src/Groups/directproducts.jl]:
```suggestion
the output is a triple (`G`, `emb`, `proj`), where `emb` and `proj` are the
```

File: src/Rings/MPolyQuo.jl [hunk 6]
```diff
@@ -871,7 +871,7 @@ end
 # The above method for `simplify` assume that there is a singular backend which 
 # can be used. However, we are using (graded) quotient rings also with coefficient 
 # rings R which can not be translated to Singular; for instance when R is again 
-# a polynomial ring, or a quotient/localization thereof, or even a `SpecOpenRing`. 
+# a polynomial ring, or a quotient/localization thereof, or even a `AffineSchemeOpenSubschemeRing`.
```
Comment (simonbrandhorst) [inline src/Rings/MPolyQuo.jl]:
```suggestion
# a polynomial ring, or a quotient/localization thereof, or even an `AffineSchemeOpenSubschemeRing`. 
```

File: src/deprecations.jl [hunk 7]
```diff
@@ -93,3 +93,40 @@ end
...
+@deprecate SpecOpen_4 AffineSchemeOpenSubscheme4
+@deprecate SpecL AffineSchemeL
+@deprecate Specs AffineSchemes
+@deprecate SpecOpens AffineSchemeOpenSubschemes
+@deprecate SpecOpenRings AffineSchemeOpenSubschemeRings
+@deprecate SpecOpenMors AffineSchemeOpenSubschemeMors
+@deprecate SpecMors AffineSchemeMors
+@deprecate AbsSpecs AbsAffineSchemes
+@deprecate _is_open_func_for_schemes_without_specopen _is_open_func_for_schemes_without_affine_scheme_open_subscheme
+@deprecate spec_open_ring_type affine_scheme_open_subscheme_ring_type
```
Comment (simonbrandhorst) [inline src/deprecations.jl]:
```suggestion
```

File: src/exports.jl [hunk 8]
```diff
@@ -1,5 +1,5 @@
 # Entries are sorted with uppercase before lowercase. To resort it,
-# execute:  sort < src/exports.jl > dummy ; mv dummy src/exports.jl
+# execute:  LC_COLLATE=C sort < src/exports.jl > dummy ; mv dummy src/exports.jl
```
Comment (simonbrandhorst) [inline src/exports.jl:2]: Good catch!

File: src/exports.jl [hunk 9]
```diff
@@ -1204,13 +1205,13 @@ export projective_geometry
 export projective_omega_group
 export projective_orthogonal_group
 export projective_plane
-export projective_scheme
 export projective_space
```
Comment (simonbrandhorst) [inline src/exports.jl]: Please keep the support for `projective_scheme`.
While `proj` is the mathematically correct name for this construction in algebraic geometry.
Oscar is for a wider audience. The name `projective_scheme` is more descriptive of what happens.
(construct a `ProjectiveScheme` from a graded ring). 
Therefore both `proj` and `projective_scheme` should be allowed.
Similarly we should have the constructors `affine_scheme` and `spec`.

Comment (lgoettgens) [reply src/exports.jl]: There are already some examples of how to achieve such an alias in Aliases.jl

**Review by simonbrandhorst — APPROVED** [review]
Thanks for the big effort Erik. I appreciate it!

File: src/exports.jl (hunk 9 above)
Comment (simonbrandhorst) [reply src/exports.jl]: Maybe someone else wants to use the abbreviations  
`proj` or `spec` in another context with a different meaning. 
Therefore I think that a global alias `proj = projective_scheme` is not justified here?
I merely added calls for `projective_scheme` with the corresponding signatures.

**Comment by paemurru** [comment]
@fingolfin I did not know that @thofma was among OSCAR project leaders, sorry about that. His name does not appear under OSCAR project leaders in the [OSCAR website](https://www.oscar-system.org/contributors/). Perhaps the page should be updated.

> You can of course request that "the OSCAR leaders" sit together and give a final verdict before you change anything, so that you don't end up changing things back and forth (a quite reasonable request).

Yes, this is what I would have liked, as it seemed to me there was no consensus. But considering how everyone is very busy with the OSCAR book and OSCAR 1.0, I didn't think it would be appropriate for me to request this

**Comment by fingolfin** [comment]
Yes, the website should be updated. We'll get to it eventually.
